### ISSUE TITLE
Support environment variable expansion in configuration files?

### ISSUE THREAD

**Issue by mikaelarguedas** [colcon/colcon-core#53]
I use environment variables in my ament aliases, for example to construct build/install destination based on the workspace I'm building and the parameters I uese to build it.

Is the recommended way to do this with colcon to create bash aliases to expand these variable and to invoke colcon or is there a plan to support such use cases in the colcon configuration system?

**Comment by dirk-thomas** [comment]
Can you post an example what you want to achieve, how you currently do it using aliases and how you imagine it to be done using configuration files expanding variables.

**Comment by mikaelarguedas** [comment]
I would like to have different build and install folders based on the workspace I'm building and environment variable.

e.g.
`cb` meaning colcon build and `cb-debug` meaning colcon build in debug mode.
`export ROS2_ws=/tmp/my_ros2_ws`
`alias cb=source $RTI_ENV_SCRIPT && pushd && cd $ROS2_WS && colcon build  --symlink-install --event-handler console_cohesion+ --cmake-args ' -DSECURITY=ON -DBUILD_TESTING=1'`
`alias cb-debug=cb ' -DCMAKE_BUILD_TYPE=Debug' --build-base $ROS2_WS/build_debug_isolated --install-base $ROS2_WS/install_debug_isolated`

I would like to push the cmake_arguments in configuration files. My default configuratioon file would set `--symlink-install --event-handler console_cohesion+ --cmake-args ' -DSECURITY=ON -DBUILD_TESTING=1'`
And my "debug_configuration" file would extend the previous defaults to apply:
`' -DCMAKE_BUILD_TYPE=Debug' --build-base $ROS2_WS/build_debug_isolated --install-base $ROS2_WS/install_debug_isolated`

This would allow me to have a generic set of configuration files regardless of where the code is located (ROS2_WS env var) and also allow me to pass cmake-args to my aliases that is currently not the case because of how argparse is used in colcon.

**Comment by dirk-thomas** [comment]
So you are using `$ROS2_WS` as the base of your workspace and use it as a prefix for the build base and install base. You are also changing into the workspace directory before. So why not just pass a relative path to `--build-base` / `--install-base`? I would expect that to work.

> `--cmake-args ' -DSECURITY=ON -DBUILD_TESTING=1'`

As mention in another ticket you should quote your CMake arguments separately: `--cmake-args ' -DSECURITY=ON' ' -DBUILD_TESTING=1'`

**Comment by mikaelarguedas** [comment]
> As mention in another ticket you should quote your CMake arguments separately: --cmake-args ' -DSECURITY=ON' ' -DBUILD_TESTING=1'

That's a typo, I quote them separately but I still can't pass another `--cmake-args` so I need to:
- hack `cb-debug` to assume cb ends with a `--cmake-args` I can add parameters to: `cb-debug=cb ' -DCMAKE_BUILD_TYPE=Debug'`
- I cannot pass additional `--cmake-args` after the alias as they overwrite the existing ones.

>  So why not just pass a relative path to --build-base / --install-base? I would expect that to work.

This was a basic example. I also have different destinations I can use to keep my source workspace clean and build/install in different locations so a relative path is not solving the problem if I want to keep my previous artifacts around.
e.g. `export ROS2_ARTIFACT_DEST=/tmp/test_feature_x`
and alias definition like:

`alias cb=source $RTI_ENV_SCRIPT && pushd && cd $ROS2_WS && colcon build --symlink-install --event-handler console_cohesion+ --cmake-args ' -DSECURITY=ON' ' -DBUILD_TESTING=1'`

`alias cb-debug=cb ' -DCMAKE_BUILD_TYPE=Debug' --build-base $ROS2_ARTIFACT_DEST/build_debug_isolated --install-base $ROS2_ARTIFACT_DEST/install_debug_isolated`

**Comment by mikaelarguedas** [comment]
Maybe it's a undesirable workflow, I was hoping to work around the `cmake-args` issue to use configuration files instead.
So I could have something like:
- `alias cb=source $RTI_ENV_SCRIPT && pushd && cd $ROS2_WS && colcon build` Taking my default configuration file
- `alias cb-debug=cb --metas $PATH_TO_MY_DEBUG_DONFIG_FILE` that would extend my defaults using the parameters from the debug config files (including the paths defined with env vars)

That should allow me to pass additional arguments without overwriting as I would pass the commandline argument only once.

But maybe I'm approaching this the wrong way. Is there another workflow that would allow me to do the same thing ?

**Comment by dirk-thomas** [comment]
This sounds very similar to #52 now. `colcon-defaults` could add an argument `--defaults` to pass multiple files and their content could be merged / concatenated and the result be used as default value.

For your two aliases you would still rely on `--defaults` being the last arg of `cb` or you would have to repeat the value in `cb-debug`.

**Comment by dirk-thomas** [comment]
You might want to try the new [mixin](https://github.com/colcon/colcon-mixin) feature for this and see how it works for your use case.

**Comment by mikaelarguedas** [comment]
I'll give mixin a shot for this and https://github.com/colcon/colcon-core/issues/52 and report back

**Comment by mikaelarguedas** [comment]
While the environment variables expansion is not present, I can work around it with a combination of aliases default and mixins so I'm going to clodse this issue.

Thanks @dirk-thomas !
